### PR TITLE
[cocoon] Remove user authentication when backend calls fail.

### PR DESCRIPTION
**Pull request by ditman**
This PR updates the Build Dashboard to make sure its UI reflects the authentication state of the user more directly. The UI refreshes as "logged out" when calls to the backend that require a valid authentication fail.

Before this, it was not entirely clear why some actions failed (failing operations only raised a small snackbar at the bottom of the page, and there was no way to renew the user session other than full-page reloading). The app could be in an inconsistent state where it *seemed* to be logged-in, but the user authorization would have expired.

After this PR, the app will seemingly "kick the user out" when they attempt something that they don't have permission to do, so they can interactively re-authenticate. It will also disable the problematic buttons so they cannot be spammed once the backend deauthorizes a user.

The changes are:

* Stops asking "Google Sign In" plugin if the current user `isSignedIn`, it has no idea.
* Makes the `isAuthenticated` check local to Cocoon, and not async (`return user != null`)
  * Updates mocks (and tests) to account for the API changes.
* Adds a `clearUser` method to refresh the UI but not tell the plugin that the user has signed out, so re-login is faster.
  * Calls `clearUser` whenever any request to the server that has `idToken` fails.
    * Assumes that requests only fail for authentication reasons, which might not always be the case! Opportunity for improvement here!
    * Currently this applies only to two requests: `refreshGitHubCommits` and `rerunTask`.
* Minimizes the chances of users sending unauthorized requests to the server.
  * Disables the "RERUN" and "Refresh GitHub Commits" buttons when the user is deemed not authenticated, so they must re-login rather than being able to click forever.
  * Removes the logic that did `signIn` if `idToken` was missing. It doesn't work as intended anymore. Instead, harden the `idToken` getter so it *requires* `isAuthenticated`.
* Adds tests for all of the above.

## Issues

* Mitigates/Fixes https://github.com/flutter/flutter/issues/125171

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read the [Flutter Style Guide] _recently_, and have followed its advice.
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Review by CaseyHillers — APPROVED** [review]
LGTM!

**Review by ricardoamador — APPROVED** [review]
LGTM! Thanks for taking care of this David!

File: dashboard/lib/state/build.dart [hunk 1]
```diff
@@ -359,12 +362,26 @@ class BuildState extends ChangeNotifier {
     assert(_statusesAreUnique(statuses));
   }
 
-  Future<bool> refreshGitHubCommits() async => cocoonService.vacuumGitHubCommits(await authService.idToken);
+  Future<bool> refreshGitHubCommits() async {
+    if (!authService.isAuthenticated) {
+      return false;
+    }
+    final bool successful = await cocoonService.vacuumGitHubCommits(await authService.idToken);
+    if (!successful) {
```
[the remark this answers is not in the document]
Comment (ditman) [reply dashboard/lib/state/build.dart:370]: The cocoon service does not return the status code (only a boolean or something-that-contains-a-boolean), this needs a little bit more love if we want to be more specific (I'd return a `(bool success, int errCode)` or similar object)

**Comment by ditman** [comment]
I need to take a look at the tests that are failing, I'd swear I fixed them in my gLinux (and maybe forgot to save/commit a file! (x_x))

File: dashboard/test/widgets/task_overlay_test.dart [hunk 2]
```diff
@@ -25,16 +27,16 @@ import '../utils/golden.dart';
 import '../utils/task_icons.dart';
 
 class TestGrid extends StatelessWidget {
-  const TestGrid({this.buildState, required this.task, super.key});
+  const TestGrid({required this.buildState, required this.task, super.key});
```
[the remark this answers is not in the document]
Comment (ditman) [reply dashboard/test/widgets/task_overlay_test.dart]: Yep, added.

File: dashboard/lib/state/build.dart (hunk 1 above)
[the remark this answers is not in the document]
Comment (ditman) [reply dashboard/lib/state/build.dart:370]: @CaseyHillers I've created this issue for now, I don't want to fix that right now (there's a looot of code testing those methods, I think)

* https://github.com/flutter/flutter/issues/125440

I can take a look at the return types in another PR so we're more surgical when logging-out the user :)

**Comment by ditman** [comment]
I've reproduced the test failures by running the tests in a random order: 

```
flutter test --test-randomize-ordering-seed=random
```

I'll try to make the tests more hermetic.

**Comment by ditman** [comment]
Let's land this, see what people think!

**Pull request merged**